### PR TITLE
Vm json test

### PR DESCRIPTION
**Pull request by shargon**
Is a draft

TODO:

Deal with comples types like: Array, Struct, Map and Interop

### CODE REVIEW

**Comment by shargon** [comment]
I have issues with NeoVM fixed here: https://github.com/neo-project/neo-vm/pull/69

File: test/NeoSharp.VM.Test/VMJsonTest.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      ],
+      ""Trigger"": ""Application""
+    }
+  ]
+}
+";
+
+            ExecuteTest(ut);
+
+            ut =
```
Comment (osmirnov) [inline test/NeoSharp.VM.Test/VMJsonTest.cs]: Maybe, create json file and reference it?

File: test/NeoSharp.VM.Test/VMJsonTest.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+@"
+{
+  ""Category"": ""PUSHES"",
```
Comment (osmirnov) [inline test/NeoSharp.VM.Test/VMJsonTest.cs]: I suggest camelCase props

File: test/NeoSharp.VM.Test/VMJsonTestBase.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        /// <param name="message">Message</param>
+        private void AssertResult(Stack stack, VMUTStackItem[] result, string message)
+        {
+            Assert.AreEqual(stack.Count, result == null ? 0 : result.Length, message + "Stack is different");
+
+            for (int x = 0, max = stack.Count; x < max; x++)
+            {
+                var item = stack.Peek(x);
+
+                Assert.AreEqual(item.Type.ToString(), result[x].Type.ToString(), message + "Stack item type is different");
```
Comment (osmirnov) [inline test/NeoSharp.VM.Test/VMJsonTestBase.cs]: It'd be helpful to say `Expected` and `Actual`. The same for all asserts.

File: test/NeoSharp.VM.Test/VMJsonTestBase.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+namespace NeoSharp.VM.Test
+{
+    public abstract class VMJsonTestBase
+    {
+        public void ExecuteTest(string json)
+        {
+            var ut = json.JsonToVMUT();
+
+            foreach (var test in ut.Tests)
```
Comment (osmirnov) [inline test/NeoSharp.VM.Test/VMJsonTestBase.cs:22]: I am questioning is it `Tests` or `Steps` of a single test?

Comment (shargon) [reply test/NeoSharp.VM.Test/VMJsonTestBase.cs:22]: Steps is better , i will change it

File: test/NeoSharp.VM.Test/VMJsonTest.cs (hunk 1 above)
Comment (shargon) [reply test/NeoSharp.VM.Test/VMJsonTest.cs]: Yes, is just a sample, will create json files

File: test/NeoSharp.VM.Test/VMJsonTestBase.cs (hunk 4 above)
Comment (shargon) [reply test/NeoSharp.VM.Test/VMJsonTestBase.cs:22]: This is tests, actions is steps :P

**Review by osmirnov — APPROVED** [review]

**Pull request merged**